components/sup/src/event/stream.rs

```diff
@@ -3,31 +3,81 @@ use crate::event::{Error,
+    // TODO (DM): This cannot be unbounded. We need backpressure. If the connection is down when we
+    // try to publish we try to reconnect this can be time consuming so we can easily get
+    // behind.
     let (event_tx, event_rx) = futures_mpsc::unbounded();
     let (sync_tx, sync_rx) = std_mpsc::sync_channel(0); // rendezvous channel
 
     let EventStreamConnectionInfo { name,
                                     verbose,
                                     cluster_uri,
-                                    cluster_id,
-                                    auth_token,
+                                    // TODO (DM): The nats client we are using does not support
+                                    // auth tokens and will need to be patched
+                                    auth_token: _,
                                     connect_method, } = conn_info;
-    let connect_method_is_timeout = connect_method.is_timeout();
+    let connection_is_timeout = connect_method.is_timeout();
+
+    // Note: With the way we are using the client, we will not respond to pings from the server
+    // (https://nats-io.github.io/docs/nats_protocol/nats-protocol.html#pingpong).
+    // Instead, we rely on publishing events to keep us connected to the server. This is completely
+    // valid according to the protocol, and in fact, the server will not send pings if there is
+    // other traffic from a client. If we have no events for an extended period of time, we will
+    // be automatically disconnected (because we do not respond to pings). When the next event comes
+    // in we will try to reconnect.
+    let mut client = Client::new(cluster_uri)?;
+    client.set_name(&name);
+    client.set_synchronous(verbose);
+    let closure = move || {
+        // Try to establish an intial connection to the NATS server.
+        loop {
+            if let Err(e) = client.connect() {
+                if connection_is_timeout {
+                    error!("Failed to connect to NATS server '{}'. Retrying...", e);
+                } else {
+                    warn!("Failed to connect to NATS server '{}'.", e);
+                    break;
+                }
+            } else {
+                if connection_is_timeout {
+                    sync_tx.send(())
+                           .expect("Couldn't synchronize event thread!");
+                }
+                break;
+            }
+            thread::sleep(Duration::from_secs(1))
+        }
+
+        let event_handler = event_rx.for_each(move |event: Vec<u8>| {
+                                        if let Err(e) = client.publish(HABITAT_SUBJECT, &event) {
+                                            error!("Failed to publish event, '{}'", e);
+                                        }
+                                        Ok(())
+                                    });
+
+        Runtime::new().expect("Couldn't create event stream runtime!")
```

Comment:
It's probably worth noting here: the reason all this was initially running in a thread in the first place is that the `nitox` library had an issue where it didn't play well with other futures on its reactor. To work around that, I just put it off on its own reactor on a separate thread.

Since `rust-nats` presumably doesn't have that issue, we _could_ theoretically move all this into running directly on the Supervisor's main reactor. If we were to do that, however, we'd need to do it in such a way that we could cleanly shut it down when the Supervisor needs to come down, or run into the same underlying issue that was behind https://github.com/habitat-sh/habitat/issues/6712 and fixed by https://github.com/habitat-sh/habitat/pull/6717.

(It's perfectly fine to leave it as is; I'm just thinking it would be good to leave a comment here to ensure that any well-meaning refactoring engineer that comes after us knows what the trade-offs are.)

Reply:
These commits [here](https://github.com/habitat-sh/habitat/pull/6853/commits/15228661d18f3ff4adcde8d5072e66934b7e941c) and [here](https://github.com/habitat-sh/habitat/pull/6853/commits/e62b6ef813b84ecc5b302431eaad2325bf0a1918) address this.

It seems like we need a more robust and ergonomic solution for ensuring all futures are stopped before shutting down. This is outside the scope of this PR, but just to get some ideas out there.

What if we made a "wrapper" that wraps two runtimes. This "wrapper" could have a `spawn_divergent` (or something to indicate that this future does not end) call. This would spawn the future on a runtime that calls `shutdown_now` instead of `shutdown_on_idle` when we shutdown.

Not sold on this solution, but it would be nice to not have to keep a handle for every divergent future. I wonder how other tokio projects handle correctly ending unbounded futures. Thoughts?

Reply:
Neat :smile: 

I'm not sure what the best way going forward should be for this. I like the "handle" approach, since it's explicit, but it does require a little bookkeeping. I haven't seen other approaches for this, though (which is what motivated that handle solution initially).

Your "two Runtimes" approach is also an interesting one, and is worth digging into, I think.

As long as the code in this PR doesn't get us back into a 0.83.0 bug situation, I'm :+1: on merging it.